qiskit/primitives/utils.py

```diff
@@ -130,8 +130,32 @@ def _circuit_key(circuit: QuantumCircuit, functional: bool = True) -> tuple:
         circuit.num_qubits,
         circuit.num_clbits,
         circuit.num_parameters,
-        tuple(
-            (d.qubits, d.clbits, d.operation.name, tuple(d.operation.params)) for d in circuit.data
+        tuple(  # circuit.data
+            (
+                tuple(  # qubits
+                    (
+                        circuit.find_bit(qubit).index,
+                        tuple(
+                            (reg[0].size, reg[0].name, reg[1])
+                            for reg in circuit.find_bit(qubit).registers
+                        ),
+                    )
+                    for qubit in data.qubits
+                ),
```

Comment:
Since the codes for qubit and clbits look identical, I suggest making a private function to avoid duplication.